pest/src/span.rs

```diff
@@ -239,30 +275,82 @@ impl<'i> Hash for Span<'i> {
 ///
 /// [`Span::lines()`]: struct.Span.html#method.lines
 pub struct Lines<'i> {
-    span: &'i Span<'i>,
-    pos: usize,
+    inner: LinesSpan<'i>,
 }
 
 impl<'i> Iterator for Lines<'i> {
     type Item = &'i str;
     fn next(&mut self) -> Option<&'i str> {
+        self.inner.next().map(|span| span.start_pos().line_of())
+    }
+}
+
+/// Like `Lines`, but returns `Span`s to preserve source mapping information. `Lines` simply calls this and `span.start_pos().line_of()`.
+///
+/// TODO better docs
+pub struct LinesSpan<'i> {
+    span: Span<'i>,
+    pos: usize,
+}
+
+impl<'i> Iterator for LinesSpan<'i> {
+    type Item = Span<'i>;
+    fn next(&mut self) -> Option<Span<'i>> {
         if self.pos > self.span.end {
             return None;
         }
         let pos = position::Position::new(self.span.input, self.pos)?;
         if pos.at_end() {
             return None;
         }
-        let line = pos.line_of();
+        let line_start = self.pos;
         self.pos = pos.find_line_end();
-        Some(line)
+        Some(unsafe {
+            Span::new_unchecked(self.span.input, line_start, self.pos.min(self.span.end))
+        })
```

Comment:
@Alextopher could this be done without `unsafe`? if not, it should have a `Safety` comment explaining why it is safe or when things could go wrong